### PR TITLE
TESTENG-1496: EnableApplicationInsights now respects a tri-state (All, UsageOnly, None)

### PR DESCRIPTION
**Pull request by matthew-wiemer**
pushing these out eod and before implementation

Complete the following tasks before merging in your PR:
- [x] If this is a user facing change, update [version.txt](https://github.com/relativitydev/relativity.testing.framework.api/blob/master/version.txt) and [CHANGELOG.md](https://github.com/relativitydev/relativity.testing.framework.api/blob/master/CHANGELOG.md)

### CODE REVIEW

**Comment by paulletang** [comment]
Anything specific you're looking for feedback on with the draft PR?

**Comment by matthew-wiemer** [comment]
> Anything specific you're looking for feedback on with the draft PR?

No, just wanted to get it up if people had feedback on missing cases or anything minor while I worked the implementation.

File: source/Relativity.Testing.Framework.Api/Interceptors/ApplicationInsightsEventInterceptor.cs [hunk 1]
```diff
@@ -13,7 +13,9 @@ public ApplicationInsightsEventInterceptor(IRelativityFacade relativityFacade)
 
 		public override void Intercept(IInvocation invocation)
 		{
+#pragma warning disable CS0618 // Type or member is obsolete
```
Comment (matthew-wiemer) [inline source/Relativity.Testing.Framework.Api/Interceptors/ApplicationInsightsEventInterceptor.cs:16]: These are in place to show everything else works without breaking the build over an obsolete usage (which should be removed in another PR)

**Comment by matthew-wiemer** [comment]
I think this is good to merge, though it doesn't change much yet. Adding in the interceptor updates will only increase the size here.

File: source/Relativity.Testing.Framework.Api.Tests/Utilities/TestApplicationInsightsInterceptor.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	{
+		public TestApplicationInsightsInterceptor(IRelativityFacade facade)
+			: base(facade)
+		{
+		}
+
+		public override void Intercept(IInvocation invocation)
+		{
+			// Nothing (yet), just need access to public properties
+			throw new NotImplementedException();
```
Comment (klaudiaknafel) [inline source/Relativity.Testing.Framework.Api.Tests/Utilities/TestApplicationInsightsInterceptor.cs:21]: shouldn't this be implemented before merging?

Comment (matthew-wiemer) [reply source/Relativity.Testing.Framework.Api.Tests/Utilities/TestApplicationInsightsInterceptor.cs:21]: No, this is implemented as much as necessary for the tests in this PR.

Comment (paulletang) [reply source/Relativity.Testing.Framework.Api.Tests/Utilities/TestApplicationInsightsInterceptor.cs:21]: Can you elaborate on why this is the case so that others have context for this?

File: source/Relativity.Testing.Framework.Api/Interceptors/ApplicationInsightsInterceptor.cs [hunk 3]
```diff
@@ -32,7 +38,37 @@ protected ApplicationInsightsInterceptor(IRelativityFacade relativityFacade)
...
+		/// <remarks>
+		/// This will return 'false' when CollectionState is 'UsageOnly', this is to avoid
+		/// any external client violating that state (by incorrectly treating configuration as 'All').
+		/// </remarks>
+		[Obsolete("This property exists to maintain any backwards compatibility; update to use CollectionState.")]
+		public bool IsEnabled
+		{
+			get
+			{
+				_enabled = false;
```
Comment (paulletang) [inline source/Relativity.Testing.Framework.Api/Interceptors/ApplicationInsightsInterceptor.cs:53]: Every time we get the `IsEnabled` property it's going to re-set the `_enabled` property?
This feels weird to me to have getter function as a setter.

Comment (paulletang) [reply source/Relativity.Testing.Framework.Api/Interceptors/ApplicationInsightsInterceptor.cs:53]: I'm still mulling this over, but it feels like less misdirection if we just store both in a normal way, and then take the lowest value provided.

File: source/Relativity.Testing.Framework.Api/Interceptors/ApplicationInsightsInterceptor.cs [hunk 4]
```diff
@@ -32,7 +38,37 @@ protected ApplicationInsightsInterceptor(IRelativityFacade relativityFacade)
...
+			set
+			{
+				_enabled = value;
+			}
+		}
+
+		/// <summary>
+		/// Gets or sets a value indicating what type of data is being collected.
+		/// </summary>
+		public DataCollection CollectionState { get; set; } = DataCollection.UsageOnly;
```
Comment (paulletang) [inline source/Relativity.Testing.Framework.Api/Interceptors/ApplicationInsightsInterceptor.cs:71]: With this as the default value, how are we going to continue to collect stacktraces from internal builds?

File: source/Relativity.Testing.Framework.Api.Tests/Interceptors/ApplicationInsightsInterceptorTests.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
+﻿using Moq;
+using NUnit.Framework;
+using Relativity.Testing.Framework.Api.Tests.Utilities;
+
+namespace Relativity.Testing.Framework.Api.Tests
+{
+	[TestFixture]
+	public class ApplicationInsightsInterceptorTests
```
Comment (paulletang) [inline source/Relativity.Testing.Framework.Api.Tests/Interceptors/ApplicationInsightsInterceptorTests.cs:8]: This should follow our naming standards for test fixtures.

File: source/Relativity.Testing.Framework.Api.Tests/Interceptors/ApplicationInsightsInterceptorTests.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+using Relativity.Testing.Framework.Api.Tests.Utilities;
+
+namespace Relativity.Testing.Framework.Api.Tests
+{
+	[TestFixture]
+	public class ApplicationInsightsInterceptorTests
+	{
+		private TestApplicationInsightsInterceptor _unit;
+
+		private Mock<IRelativityFacade> _mockFacade; // Ignored.
```
Comment (paulletang) [inline source/Relativity.Testing.Framework.Api.Tests/Interceptors/ApplicationInsightsInterceptorTests.cs:12]: I don't know what this comment is telling me.
What does it mean for a Facade to be ignored?

File: source/Relativity.Testing.Framework.Api.Tests/Interceptors/ApplicationInsightsInterceptorTests.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+			_mockFacade
+				.Setup(x => x.Resolve<IApplicationInsightsTelemetryClient>())
+				.Returns(_mockApplicationInsightsClient.Object);
+
+			_unit = new TestApplicationInsightsInterceptor(_mockFacade.Object);
+		}
+
+		[Test]
+		public void Ensure_OldBehavior_MatchesNewBehavior([Values] DataCollection testCollectionState)
```
Comment (paulletang) [inline source/Relativity.Testing.Framework.Api.Tests/Interceptors/ApplicationInsightsInterceptorTests.cs:29]: Without the context for `OldBehavior` and `NewBehavior`, how will someone looking at this test know what it is supposed to be doing?

File: source/Relativity.Testing.Framework.Api.Tests/ApiComponent/ApiComponentApplicationInsightsTests.cs [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+		// Indirect mocks (resolved by container; resolves 'EnableApplicationInsights')
+		private Mock<IConfigurationService> _mockConfigurationService;
+
+		// Ignored mocks (resolved/required, but irrelevant to this suite)
+		private Mock<IRelativityFacade> _mockFacade;
+		private Mock<IConfigurationStore> _mockConfigurationStore;
+		private Mock<IApplicationInsightsTelemetryClient> _mockApplicationInsightsClient;
+
+		[SetUp]
+		public void SetUp()
```
Comment (paulletang) [inline source/Relativity.Testing.Framework.Api.Tests/ApiComponent/ApiComponentApplicationInsightsTests.cs:33]: Does this need to be set up for every test, or should this be a `OneTimeSetUp`?

File: source/Relativity.Testing.Framework.Api.Tests/Interceptors/ApplicationInsightsInterceptorTests.cs (hunk 5 above)
Comment (matthew-wiemer) [reply source/Relativity.Testing.Framework.Api.Tests/Interceptors/ApplicationInsightsInterceptorTests.cs:8]: - [ ] tests -> fixtures.

File: source/Relativity.Testing.Framework.Api.Tests/Interceptors/ApplicationInsightsInterceptorTests.cs (hunk 6 above)
Comment (matthew-wiemer) [reply source/Relativity.Testing.Framework.Api.Tests/Interceptors/ApplicationInsightsInterceptorTests.cs:12]: Will update. I mainly use these to remind myself that an implementation forced me to set this up, but I don't actually need or am interested in it in the context of the test suite. Doesn't really fit here and I think it's just a crumb I should clean up.

File: source/Relativity.Testing.Framework.Api.Tests/Interceptors/ApplicationInsightsInterceptorTests.cs (hunk 7 above)
Comment (matthew-wiemer) [reply source/Relativity.Testing.Framework.Api.Tests/Interceptors/ApplicationInsightsInterceptorTests.cs:29]: Will clean this up.

File: source/Relativity.Testing.Framework.Api.Tests/Utilities/TestApplicationInsightsInterceptor.cs (hunk 2 above)
Comment (matthew-wiemer) [reply source/Relativity.Testing.Framework.Api.Tests/Utilities/TestApplicationInsightsInterceptor.cs:21]: My bad - I'll add more commentary here in the code.

The existing tests only concern themselves with the base class's `IsEnabled` property. Nothing being tested here requires an implementation of `Intercept` (and if it did, it would like be a test of the relevant Interceptor implementation.)

The only (debatable) reason this class exists at all is to avoid using a production interceptor in a test class (to exercise the public members of the base class)

File: source/Relativity.Testing.Framework.Api/ApiComponent.cs [hunk 9]
```diff
@@ -70,10 +70,36 @@ void IWindsorInstaller.Install(IWindsorContainer container, IConfigurationStore
...
 			var configurationService = container.ResolveAll<IConfigurationService>().FirstOrDefault();
 
-			if (configurationService != null && !configurationService.GetValueOrDefault("EnableApplicationInsights", true))
+			if (configurationService != null)
 			{
-				container.Resolve<ApplicationInsightsEventInterceptor>().IsEnabled = false;
-				container.Resolve<ApplicationInsightsMetricInterceptor>().IsEnabled = false;
+				DataCollection stateToConfigure;
+				string defaultValue = DataCollection.UsageOnly.ToString();
+				string configuredValue = configurationService.GetValueOrDefault("EnableApplicationInsights", defaultValue).ToLower();
```
Comment (matthew-wiemer) [inline source/Relativity.Testing.Framework.Api/ApiComponent.cs:77]: Just want to revisit this before merging: do we want to change the default behavior of "missing config key" from `False` to `UsageOnly`?

Comment (briantoms) [reply source/Relativity.Testing.Framework.Api/ApiComponent.cs:77]: Wait, what? I was 99% sure the old default was `true`.

File: source/Relativity.Testing.Framework.Api/Interceptors/ApplicationInsightsInterceptor.cs (hunk 3 above)
Comment (matthew-wiemer) [reply source/Relativity.Testing.Framework.Api/Interceptors/ApplicationInsightsInterceptor.cs]: Changing this has implications for the test cases we are driving from the perspective of "IsEnabled respects DataCollection now, not the other way around"

i.e. if I remove that first set to false:

image

updated to the simpler DataCollection == All, but now we're ignoring set altogether (which I _think_ is fine)

File: source/Relativity.Testing.Framework.Api/Interceptors/ApplicationInsightsInterceptor.cs (hunk 4 above)
Comment (matthew-wiemer) [reply source/Relativity.Testing.Framework.Api/Interceptors/ApplicationInsightsInterceptor.cs:68]: I figured using the configuration key and making sure it was set was what the other task was about: https://jira.kcura.com/browse/TESTENG-1498

My recall is that we wanted to use UsageOnly as the default, but that's different from False (and neither of those force consumers to send us stack traces)

File: source/Relativity.Testing.Framework.Api/ApiComponent.cs (hunk 9 above)
Comment (matthew-wiemer) [reply source/Relativity.Testing.Framework.Api/ApiComponent.cs:77]: Yeah my mistake. From this code it looks like it would be left `False` but the base interceptor defaults the boolean to `True`.

The main question stands, iirc from planning (and one of the next up tasks discusses the default being UsageOnly). Is that what we want? Seems like a missing AC on the story.

File: source/Relativity.Testing.Framework.Api.Tests/ApiComponent/ApiComponentApplicationInsightsTests.cs (hunk 8 above)
Comment (matthew-wiemer) [reply source/Relativity.Testing.Framework.Api.Tests/ApiComponent/ApiComponentApplicationInsightsTests.cs]: Sorry, thought I posted a reply here already. I personally think it's immaterial in a unit test context, unless something is showing that it's taking too long. In my eyes it creates an un-necessary risk of the tests becoming coupled to one another/mis-managing state.

**Review by klaudiaknafel — APPROVED** [review]

**Review by klaudiaknafel — APPROVED** [review]

File: source/Relativity.Testing.Framework.Api.Tests/ApiComponent/ApiComponentApplicationInsightsTests.cs (hunk 8 above)
Comment (paulletang) [reply source/Relativity.Testing.Framework.Api.Tests/ApiComponent/ApiComponentApplicationInsightsTests.cs]: Wouldn't they be more coupled to each other if they needed a `SetUp` instead of a `OneTimeSetUp`?
